pkg/build/builder/docker.go

```diff
@@ -171,6 +176,136 @@ func (d *DockerBuilder) addBuildParameters(dir string) error {
 	return nil
 }
 
+// invalidCmdErr represents an error returned from replaceValidCmd
+// when an invalid Dockerfile command has been passed to
+// replaceValidCmd
+var invalidCmdErr = errors.New("invalid Dockerfile command")
+
+// replaceCmdErr represents an error returned from replaceValidCmd
+// when a command which has more than one valid occurrences inside
+// a Dockerfile has been passed or the specified command cannot
+// be found
+var replaceCmdErr = errors.New("cannot replace given Dockerfile command")
+
+// replaceValidCmd replaces the valid occurrence of a command
+// in a Dockerfile with the given replaceArgs
+func replaceValidCmd(cmd, replaceArgs string, fileData []byte) (string, error) {
+	if _, ok := dockercmd.Commands[cmd]; !ok {
+		return "", invalidCmdErr
+	}
+	buf := bytes.NewBuffer(fileData)
+	// Parse with Docker parser
+	node, err := parser.Parse(buf)
+	if err != nil {
+		return "", errors.New("cannot parse Dockerfile")
+	}
+
+	pos := traverseAST(cmd, node)
+	if pos == 0 {
+		return "", replaceCmdErr
+	}
+
+	// Re-initialize the buffer
+	buf = bytes.NewBuffer(fileData)
+	var newFileData string
+	var index int
+	var replaceNextLn bool
+	for {
+		line, err := buf.ReadString('\n')
+		if err != nil && err != io.EOF {
+			return "", err
+		}
+		line = strings.TrimSpace(line)
+
+		// The current line starts with the specified command (cmd)
+		if strings.HasPrefix(strings.ToUpper(line), strings.ToUpper(cmd)) {
+			index++
+
+			// The current line finishes on a backslash.
+			// All we need to do is replace the next line
+			// with our specified replaceArgs
+			if line[len(line)-1:] == "\\" && index == pos {
+				replaceNextLn = true
+
+				args := strings.Split(line, " ")
+				if len(args) > 2 {
+					// Keep just our Dockerfile command and the backslash
+					newFileData += args[0] + " \\" + "\n"
+				} else {
+					newFileData += line + "\n"
+				}
+				continue
+			}
+
+			// Normal ending line
+			if index == pos {
+				line = fmt.Sprintf("%s %s", strings.ToUpper(cmd), replaceArgs)
+			}
+		}
+
+		// Previous line ended on a backslash
+		// This line contains command arguments
+		if replaceNextLn {
+			if line[len(line)-1:] == "\\" {
+				// Ignore all successive lines terminating on a backslash
+				// since they all are going to be replaced by replaceArgs
+				continue
+			}
+			replaceNextLn = false
+			line = replaceArgs
+		}
+
+		if err == io.EOF {
+			// Otherwise, the new Dockerfile will have one newline
+			// more in the end
+			newFileData += line
+			break
+		}
+		newFileData += line + "\n"
+	}
+
+	// Parse output for validation
+	buf = bytes.NewBuffer([]byte(newFileData))
+	if _, err := parser.Parse(buf); err != nil {
+		return "", errors.New("cannot parse new Dockerfile")
```

Comment:
include the error you got parsing the new file, it'll (hopefully) be useful to debug how the file is corrupt after replacement.